database/sql/src/main/scala/cromwell/database/slick/tables/WorkflowStoreEntryComponent.scala

```diff
@@ -73,31 +73,96 @@ trait WorkflowStoreEntryComponent {
   )
 
   /**
-    * Returns up to "limit" startable workflows, sorted by submission time.
-    */
-  def fetchStartableWorkflows(limit: Long,
-     heartbeatTimestampTimedOut: Timestamp,
-     excludeWorkflowState: String,
-     excludedGroups: Set[String]
-    ): Query[WorkflowStoreEntries, WorkflowStoreEntry, Seq] = {
-      val query = for {
-        row <- workflowStoreEntries
-        /*
+   * Return hog group with the lowest count of actively running workflows
+   */
+  def getHogGroupWithLowestRunningWfs(heartbeatTimestampTimedOut: Timestamp,
+                                      excludeWorkflowState: String,
+                                      excludedGroups: Set[String]): Query[Rep[Option[String]], Option[String], Seq] = {
+    val startableWorkflows = for {
+      row <- workflowStoreEntries
+      /*
         This looks for:
-
         1) Workflows with no heartbeat (newly submitted or from a cleanly shut down Cromwell).
         2) Workflows with old heartbeats, presumably abandoned by a defunct Cromwell.
+        3) Workflows not in "OnHold" state
+        4) Workflows that don't belong to hog groups in excludedGroups
+      */
+      if (row.heartbeatTimestamp.isEmpty || row.heartbeatTimestamp < heartbeatTimestampTimedOut) &&
+        (row.workflowState =!= excludeWorkflowState) &&
+        !(row.hogGroup inSet excludedGroups)
+    } yield row
+
+    // calculates the count of startable workflows per hog group and oldest submission timestamp of workflow
+    // present in that hog group
+    val numOfStartableWfsByHogGroup = startableWorkflows
+      .groupBy(_.hogGroup)
+      .map { case (hogGroupName, groups) => (hogGroupName, groups.length, groups.map(_.submissionTime).min) }
+      .sortBy(_._2.asc)
+
+    val totalWorkflows = for {
+      row <- workflowStoreEntries
+      /*
+        This looks for:
+        1) Workflows not in "OnHold" state
+        2) Workflows that don't belong to hog groups in excludedGroups
+      */
+      if row.workflowState =!= excludeWorkflowState &&
+        !(row.hogGroup inSet excludedGroups)
+    } yield row
+
+    // calculates the count of total workflows per hog group
+    val totalWorkflowsByHogGroup = totalWorkflows
+      .groupBy(_.hogGroup)
+      .map { case (hogGroupName, groups) => (hogGroupName, groups.length) }
+      .sortBy(_._2.asc)
+
+    // calculates the number of actively running workflows for each hog group. If a hog group
+    // has all it's workflows that are either actively running or in "OnHold" status it is not
+    // included in the list. Hog groups that have no workflows actively running return count as 0
+    val wfsRunningPerHogGroup = for {
+      (t_group, t_ct) <- totalWorkflowsByHogGroup
+      (s_group, s_ct, s_sub_time) <- numOfStartableWfsByHogGroup if t_group === s_group
```

Comment:
Any change of slightly more meaningful names?

Reply:
I will update them to more meaningful names 👍